# GQL022: AOT schema attributes must be applied to classes deriving from AotSchema

|                        | Value   |
| ---------------------- | ------- |
| **Rule ID**            | GQL022  |
| **Category**           | Usage   |
| **Default severity**   | Error   |
| **Enabled by default** | Yes     |
| **Code fix provided**  | No      |
| **Introduced in**      | v9.0    |

## Cause

This rule is triggered when an attribute derived from `AotSchemaAttribute` (such as
`AotInputType`, `AotOutputType`, `AotGraphType`, `AotTypeMapping`, `AotQueryType`,
`AotMutationType`, or `AotSubscriptionType`) is applied to a class that does not
derive from `AotSchema`.

## Rule description

AOT (Ahead-of-Time) schema attributes are designed to configure schema generation
for classes that derive from `AotSchema`. These attributes provide metadata to a
source generator that produces GraphQL type definitions at compile time. Applying
these attributes to classes that don't derive from `AotSchema` serves no purpose
and indicates a configuration error.

## How to fix violations

Ensure the class with AOT schema attributes derives from `AotSchema`, or remove
the AOT schema attributes if they were applied incorrectly.

## Example of a violation

### Applying attribute to a regular class

```c#
using GraphQL;

namespace Sample;

[AotInputType(typeof(MyModel))]  // Error: MyClass doesn't derive from AotSchema
public class MyClass
{
}
```

### Applying attribute to a Schema class

```c#
using GraphQL;
using GraphQL.Types;

namespace Sample;

[AotQueryType(typeof(Query))]  // Error: MySchema derives from Schema, not AotSchema
public class MySchema : Schema
{
}
```

## Example of how to fix

### Change base class to AotSchema

```c#
using GraphQL;
using GraphQL.Types;

namespace Sample;

[AotInputType(typeof(MyModel))]
public partial class MySchema : AotSchema  // Fixed: Now derives from AotSchema
{
}
```

### Remove attribute if applied incorrectly

```c#
using GraphQL;

namespace Sample;

// Fixed: Removed the attribute since this is not a schema class
public class MyClass
{
}
```

## Related attributes

All of the following attributes derive from `AotSchemaAttribute` and must be
applied only to classes deriving from `AotSchema`:

- `AotInputTypeAttribute` - Specifies CLR types to generate InputObjectGraphType for
- `AotOutputTypeAttribute` - Specifies CLR types to generate ObjectGraphType for
- `AotGraphTypeAttribute` - Registers hand-written graph types
- `AotTypeMappingAttribute` - Maps CLR types to specific GraphQL types
- `AotQueryTypeAttribute` - Configures the query root type
- `AotMutationTypeAttribute` - Configures the mutation root type
- `AotSubscriptionTypeAttribute` - Configures the subscription root type

## Suppress a warning

If you just want to suppress a single violation, add preprocessor directives to
your source file to disable and then re-enable the rule.

```csharp
#pragma warning disable GQL022
// The code that's violating the rule is on this line.
#pragma warning restore GQL022
```

To disable the rule for a file, folder, or project, set its severity to `none` in the
[configuration file](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files).

```ini
[*.cs]
dotnet_diagnostic.GQL022.severity = none
```

For more information, see
[How to suppress code analysis warnings](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/suppress-warnings).
